### PR TITLE
Convert single loss item to list in torch estimators

### PR DESCRIPTION
**Pull request by abditag2**

### CODE REVIEW

File: horovod/spark/torch/remote.py [hunk 1]
```diff
@@ -36,9 +36,12 @@
 
 
 def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_idx):
+    # Utility functions
+    make_list_if_not_list = make_list_if_not_list_fn()
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: Since we're using this function before entering the remote code, no need to make it a generator function.

File: horovod/spark/torch/remote.py [hunk 2]
```diff
@@ -548,3 +550,15 @@ def calculate_loss(outputs, labels, loss_weights, loss_fns, sample_weights=None)
         return loss
 
     return calculate_loss
+
+
+def make_list_if_not_list_fn():
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: I would move this to `horovod/spark/common/util.py`.

File: test/test_spark_torch.py [hunk 3]
```diff
@@ -354,3 +354,20 @@ def test_torch_direct_parquet_train(self):
                     transformer = est.fit_on_parquet()
                     predictions = transformer.transform(df)
                     assert predictions.count() == df.count()
+
+    def test_make_list_if_not_list(self):
```
Comment (tgaddair) [inline test/test_spark_torch.py]: If we move the function then this should be in `test_spark.py`.

File: horovod/spark/torch/remote.py [hunk 4]
```diff
@@ -36,9 +36,12 @@
 
 
 def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_idx):
+    # Utility functions
+    make_list_if_not_list = make_list_if_not_list_fn()
+
     # Estimator parameters
-    loss_fns_pre_train = estimator.getLoss()
-    loss_constructors = estimator.getLossConstructors()
+    loss_fns_pre_train = make_list_if_not_list(estimator.getLoss())
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: What if you have multiple label cols but only want to specify a single loss.  Do we need to ensure that the list length is equal to the number of label cols?

Comment (tgaddair) [inline horovod/spark/torch/remote.py]: We have tests in `test_spark_torch.py` that do not provide loss as a list, but the tests are passing.  Based on this, I'm not sure how that's possible, do you?

Comment (abditag2) [reply horovod/spark/torch/remote.py]: If we specify a single loss for a multi-output model, that loss will be used for all the outputs.

File: horovod/spark/common/params.py [hunk 5]
```diff
@@ -156,6 +156,8 @@ def getStore(self):
         return self.getOrDefault(self.store)
 
     def setLoss(self, value):
+        if not isinstance(value, list):
```
Comment (tgaddair) [inline horovod/spark/common/params.py]: I'm not a fan of this solution because it results in counter-intuitive behavior:

```
estimator.setLoss(loss)
estimator.getLoss() == loss  # False, but expected True
```

We should do the wrapping etc. only at train time.

File: horovod/spark/torch/remote.py [hunk 6]
```diff
@@ -56,13 +56,12 @@ def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_id
...
 
     # If loss weight is not provided, use equal loss for all the labels
-    label_columns = estimator.getLabelCols()
     loss_weights = estimator.getLossWeights()
     if not loss_weights:
         num_labels = len(label_columns)
         loss_weights = [float(1) / num_labels for _ in range(num_labels)]
 
-    # Utility functions
+    # More utility functions
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: Just `Utility functions`.

File: horovod/spark/common/params.py (hunk 5 above)
Comment (abditag2) [reply horovod/spark/common/params.py]: I see your point, will revert back.

File: horovod/spark/torch/remote.py (hunk 1 above)
Comment (abditag2) [reply horovod/spark/torch/remote.py]: fixed.

File: test/test_spark_torch.py (hunk 3 above)
Comment (abditag2) [reply test/test_spark_torch.py]: moved.

File: horovod/spark/torch/remote.py (hunk 2 above)
Comment (abditag2) [reply horovod/spark/torch/remote.py]: moved.

File: horovod/spark/torch/remote.py (hunk 4 above)
Comment (abditag2) [reply horovod/spark/torch/remote.py]: In the init function of the TorchEstimator, we had a check if loss and loss_constructors were a list, and if not, we made them a list. This problem came up when we use est.setLoss().

File: horovod/spark/torch/remote.py (hunk 6 above)
Comment (abditag2) [reply horovod/spark/torch/remote.py]: NA anymore.

File: test/test_spark_torch.py [hunk 7]
```diff
@@ -343,14 +343,16 @@ def test_torch_direct_parquet_train(self):
...
                         optimizer=optimizer,
-                        loss=loss,
                         input_shapes=[[2]],
                         feature_cols=['features'],
                         label_cols=['y'],
                         batch_size=1,
                         epochs=3,
                         verbose=2)
 
+                    # To make sure that setLoss works with non-list loss.
```
Comment (tgaddair) [inline test/test_spark_torch.py]: Looks like code paths are the same for constructor and this function now, right?  Was the issue before that users were calling `setLoss` which did not coerce the param into a list?

File: test/test_spark.py [hunk 8]
```diff
@@ -772,3 +772,19 @@ def fn():
         with spark_session('test_get_available_devices', gpus=2):
             res = horovod.spark.run(fn, env={'PATH': os.environ.get('PATH')}, verbose=0)
             self.assertListEqual([(['0'], 0), (['1'], 1)], res)
+
+
+    def test_make_list_if_not_list(self):
```
Comment (tgaddair) [inline test/test_spark.py]: `test_to_list`

File: test/test_spark_torch.py (hunk 7 above)
Comment (abditag2) [reply test/test_spark_torch.py]: I believe so. Passing the loss directly during instantiation works fine as we had it in the tests.

Comment (tgaddair) [reply test/test_spark_torch.py]: I think we can move this back to the constructor, then.

File: test/test_spark_torch.py [hunk 9]
```diff
@@ -343,14 +343,16 @@ def test_torch_direct_parquet_train(self):
...
+                    est.setLoss(loss)
+
```
Comment (tgaddair) [inline test/test_spark_torch.py]: Can you add tests for `_calculate_loss_fn` in `remote.py`?  I think this problem could have been caught before if we had testing for multi-label scenarios.

File: horovod/spark/torch/remote.py [hunk 10]
```diff
@@ -56,8 +56,7 @@ def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_id
     train_minibatch = train_minibatch_fn if train_minibatch_fn else _train_minibatch_fn()
 
     # If loss weight is not provided, use equal loss for all the labels
-    label_columns = estimator.getLabelCols()
-    loss_weights = estimator.getLossWeights()
+    loss_weights = to_list(estimator.getLossWeights())
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: I don't think this will work.  If you look in https://github.com/horovod/horovod/blob/master/horovod/spark/torch/remote.py#L536 you'll see that we expect the length of loss_weights to always match the length of the labels.  We should so something similar to what we do on line 62 here.

File: horovod/spark/torch/remote.py [hunk 11]
```diff
@@ -37,8 +37,8 @@
 
 def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_idx):
     # Estimator parameters
-    loss_fns_pre_train = estimator.getLoss()
-    loss_constructors = estimator.getLossConstructors()
+    loss_fns_pre_train = to_list(estimator.getLoss())
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: In https://github.com/horovod/horovod/blob/master/horovod/spark/torch/remote.py#L522 we pad the list of loss fns on every call to `calculate_loss`.  Why not do it here?  We could use `num_labels` to ensure the loss / loss_constructors / loss_weights are the right length.  We could even add it to `to_list`:

```
loss_fns_pre_train = to_list(estimator.getLoss(), length=num_labels)
```

File: horovod/spark/common/util.py [hunk 12]
```diff
@@ -587,3 +587,23 @@ def get_dataset_properties(dataset_idx):
...
+
+
+def to_list(var, length):
+    if var is None:
+        return None
+
+    if not isinstance(var, list):
+        var = [var]
+
+    # If only one loss function or loss_constructor is passed by user, use
```
Comment (tgaddair) [inline horovod/spark/common/util.py]: Since this is in `common`, we should not reference loss and loss_constructors.

File: horovod/spark/torch/remote.py [hunk 13]
```diff
@@ -54,13 +52,21 @@ def RemoteTrainer(estimator, metadata, last_checkpoint_state, run_id, dataset_id
     user_verbose = estimator.getVerbose()
     train_minibatch_fn = estimator.getTrainMinibatchFn()
     train_minibatch = train_minibatch_fn if train_minibatch_fn else _train_minibatch_fn()
+    loss_fns_pre_train = to_list(estimator.getLoss(),
```
Comment (tgaddair) [inline horovod/spark/torch/remote.py]: Let's pull `num_labels` out from line 63 and put it just above line 55 (`loss_fns_pre_train = ...`).  That way, we can just reference `num_labels` instead of `len(label_columns)`.

File: test/test_spark_torch.py (hunk 9 above)
Comment (abditag2) [reply test/test_spark_torch.py]: added.

File: horovod/spark/torch/remote.py (hunk 11 above)
Comment (abditag2) [reply horovod/spark/torch/remote.py]: Moved it.

File: horovod/spark/common/util.py (hunk 12 above)
Comment (abditag2) [reply horovod/spark/common/util.py]: addressed
